src/devices/KeyMatrix/KeyMatrix.cs

```diff
@@ -0,0 +1,211 @@
+        {
+            _masterGpioController = masterController ?? throw new ArgumentNullException(nameof(masterController));
+
+            if (outputPins == null)
+            {
+                throw new ArgumentNullException(nameof(outputPins));
+            }
+            if (!outputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(outputPins), "The number of outputs is at least 1");
+            }
+
+            if (inputPins == null)
+            {
+                throw new ArgumentNullException(nameof(inputPins));
+            }
+            if (!inputPins.Any())
+            {
+                throw new ArgumentOutOfRangeException(nameof(inputPins), "The number of inputs is at least 1");
+            }
+
+            if (scanInterval <= 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(scanInterval), "Scanning interval must be positive");
+            }
+
+            _outputPins = outputPins.ToArray();
+            _inputPins = inputPins.ToArray();
+            _buttonValues = new PinValue[_outputPins.Length * _inputPins.Length];
+            _pinsOpened = false;
+
+            ScanInterval = scanInterval;
+
+            OpenPins();
+        }
+
+        /// <summary>
+        /// Start a task to open GPIO pins then start keyboard scanning. Raises KeyMatrixEventArgs.
+        /// </summary>
+        /// <param name="token">A cancellation token that can be used to cancel the work</param>
+        public Task ScanAsync(CancellationToken token)
+        {
+            return Task.Run(() =>
+            {
+                KeyMatrixEventArgs args = Scan(token, false);
+                return args;
+            }, token);
+        }
+
+        /// <summary>
+        /// Start a task to open GPIO pins then start keyboard scanning. End with a KeyMatrixEventArgs raises.
+        /// </summary>
+        /// <param name="token">A cancellation token that can be used to cancel the work</param>
+        public Task<KeyMatrixEventArgs> ReadKeyAsync(CancellationToken token)
+        {
+            return Task.Run(() =>
+            {
+                KeyMatrixEventArgs args = Scan(token, true);
+                return args;
+            }, token);
+        }
+
+        /// <summary>
+        /// Get buttons' values by output
+        /// </summary>
+        /// <param name="output">Output index</param>
+        public ReadOnlySpan<PinValue> ValuesByOutput(int output)
+        {
+            return _buttonValues.AsSpan(output * _outputPins.Length, _inputPins.Length);
+        }
+
+        /// <summary>
+        /// Keyboard event
+        /// </summary>
+        public event PinChangeEventHandler PinChangeEvent;
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            ClosePins();
+
+            _masterGpioController.Dispose();
```

Comment:
this should be `_masterGpioController?.Dispose();` considering you're nulling it out